### PR TITLE
Updates OSut gem files

### PR DESCRIPTION
**Pull request by brgix**
Updates/tests updates to _OSut_ gem files (see _OSut_ [PR](https://github.com/rd2/osut/pull/29)).

### CODE REVIEW

File: spec/tbd_tests_spec.rb [hunk 1]
```diff
@@ -1699,7 +1699,8 @@
     #
     #   Uo = 0.277 - ( ∑psi • L )/A
     #
-    # The method exits with an ERROR in 2x cases:
+    # If impossible to meet the requested Ut, TBD hardsets Uo to UMIN while
+    # raising warnings, namely when:
     #   - calculated Uo is negative, i.e. ( ∑psi • L )/A > 0.277
     #   - calculated layer r violates E+ material constraints, e.g.
```
Comment (brgix) [inline spec/tbd_tests_spec.rb:1741]: To trigger OpenStudio/EnergyPlus material thickness/conductivity thresholds, one must request unreasonably low U-factors (Uo < 0.010 W/m2.K, i.e. R-ip > 568). This may happen when attempting to meet NECB (2017 through 2025) prescriptive requirements, through _uprating_.

The previous version would bail out (with an error message) when hitting such thresholds, leaving the construction unaltered. The revised solution raises a big red warning, yet resets the construction Uo to UMIN (0.010 W/m2.K).

File: spec/tbd_tests_spec.rb [hunk 2]
```diff
@@ -3292,7 +3293,7 @@
     end
   end
 
-  it "can check for balcony sills (ASHRAE 90.1 2022)" do
+  it "can check for balcony sills (ASHRAE 90.1 2022/25)" do
```
Comment (brgix) [inline spec/tbd_tests_spec.rb:3510]: Not seeing any changes between 2022 vs 2025 ASHRAE 90.1 editions.

File: spec/tbd_tests_spec.rb [hunk 3]
```diff
@@ -3409,32 +3410,26 @@
...
     #   - corners
     #
-    # There are (possibly) 2x admissible interpretations of how to treat
-    # non-regulated heat losss (edges as linear thermal bridges) in 90.1:
-    #   1. assign 0 W/K•m for both proposed design and budget building models
-    #   2. assign more realistic PSI-factors, equally to both proposed/budget
+    # ... fall under the scope of requirement 5.5.5.5. There is much uncertainty
+    # on how to model items falling under 5.5.5.5. For this reason, the 8x
+    # TBD-built-in ASHRAE PSI sets have 0 W/K per meter assigned for edges under
+    # 5.5.5.5. This is discussed here:
```
Comment (brgix) [inline spec/tbd_tests_spec.rb:3640]: The only issue with thermal bridging in ASHRAE 90.1 2022/25 is requirement 5.5.5.5, discussed [here](https://unmethours.com/question/97085/901-2022-requirements-for-linear-thermal-bridges/).

File: spec/tbd_tests_spec.rb [hunk 4]
```diff
@@ -3409,32 +3410,26 @@
...
-    # treatment of (non-regulated) miscellaneous (process) loads. So for now,
-    # TBD's built-in ASHRAE 90.1 2022 (A10) PSI-factor sets recflect option #1.
+    # unmethours.com/question/97085/901-2022-requirements-for-linear-thermal-bridges
     #
-    # Users who choose option #2 can always write up a custom ASHRAE 90.1 (A10)
-    # PSI-factor set on file (tbd.json), initially based on the built-in 90.1
-    # sets while resetting non-zero PSI-factors.
+    # Users can always write up a custom ASHRAE 90.1 (A10) PSI-factor set on
+    # file (tbd.json), initially based on the built-in 90.1 sets while resetting
+    # non-zero PSI-factors.
```
Comment (brgix) [inline spec/tbd_tests_spec.rb:3646]: One can still rely on TBD for linear/point thermal bridges falling under ASHRAE 90.1 requirement 5.5.5.5, via a custom JSON file. Suggestions are also detailed [here](https://unmethours.com/question/97085/901-2022-requirements-for-linear-thermal-bridges/?answer=101707#post-id-101707).

File: lib/tbd/ua.rb [hunk 5]
```diff
@@ -185,230 +182,211 @@ def uprate(model = nil, s = {}, argh = {})
...
+          col[:s].values.each do |item|
+            col[:hloss] += item[:h]
+            col[:area ] += item[:a]
+            col[:fA   ] += item[:a] / item[:f]
+          end
 
-            if sss.isConstructionDefaulted
-              set = defaultConstructionSet(sss) # building? story?
+          # Area-weighted surface air film resistances.
+          col[:film] = 1 / ( col[:fA] / col[:area] )
```
Comment (brgix) [inline lib/tbd/ua.rb]: Other than input validation changes and minor fixes, the key change is having the (uprated) _uo_ method rely on a surface area weighted average of applicable surface air film resistances. Helps when dealing with:

- surfaces with different tilts
- (uprated) constructions referenced by both outdoor-facing and _interzone_ surfaces

File: spec/tbd_tests_spec.rb [hunk 6]
```diff
@@ -1731,14 +1730,19 @@
...
+    expect(argh).to have_key(:wall_uo)
+    expect(argh).to have_key(:roof_uo)
+    expect(argh[:wall_uo]).to_not be_nil
     expect(argh[:roof_uo]).to_not be_nil
-    expect(argh[:roof_uo]).to be_within(TOL).of(0.118) # RSi 8.47 (R48)
+
+    # Although the roof construction is correctly uprated, it is not possible to
+    # completely uprate the wall construction. It is therefore capped at the
+    # minimum allowed Uo-factor, or ~9 ft of XPS insulation.
+    expect(argh[:wall_uo].round(3)).to eq(0.010) # RSi 100.00 (R568)
```
Comment (brgix) [inline spec/tbd_tests_spec.rb]: Expected outcome when failing to _uprate_ material thickness/conductivity, e.g.:

- WARNING message: "Unable to completely uprate ... "
- uprated construction hard set at UMIN (0.010 W/m2.K), or R-ip 568

File: lib/measures/tbd/resources/utils.rb [hunk 7]
```diff
@@ -199,6 +199,61 @@ module OSut
...
+
+      if type == :ceiling || type == :partition
+        # Interzone. Fetch reciprocal tilt, e.g. if tilt == 0°, tiltx = 180°
+        tiltx = tilt + Math::PI
+
+        # Assuming tilt is contrained [0°, 180°] - constrain tiltx [0° 180°]:
+        #   e.g. tiltx == 210° if tilt ==  30°, so convert tiltx to 150°
+        #   e.g. tiltx == 330° if tilt == 150°, so convert tiltx to  30°
+        #   e.g. tiltx == 275° if tilt ==  95°, so convert tiltx to  85°
+        tiltx = Math::PI - tilt if tiltx > Math::PI
```
Comment (brgix) [inline lib/measures/tbd/resources/utils.rb:246]: New OSut _filmResistances_ method (see [PR](https://github.com/rd2/osut/pull/29)), more aligned with EnergyPlus reported values when dealing with non-vertical _interzone_ paired surfaces.

File: lib/tbd/ua.rb [hunk 8]
```diff
@@ -128,7 +128,7 @@ def uo(id = "", lc = nil, area = 0, film = 0, hloss = 0, ut = 0)
     h = format "%.3f", loss
     log(WRN, "Can't set #{h} W/K to #{id} #{mth}") if loss > TOL
 
-    u
+    uo
```
Comment (brgix) [inline lib/tbd/ua.rb:134]: Nasty oversight at this stage. Fixed.

File: spec/tbd_tests_spec.rb [hunk 9]
```diff
@@ -1945,6 +1944,184 @@
     ut = uA / m2
     expect((uA/m2).round(2)).to eq(argh[:wall_ut].round(2))
     expect(nb).to eq(4)
+
+    # --- --- --- --- --- --- --- --- --- --- --- --- --- --- --- --- --- --- #
+    # Side test, simple model:
+    #   - 3x roofceiling surfaces
+    #   - different tilts
+    #   - different outside boundary conditions
```
Comment (brgix) [inline spec/tbd_tests_spec.rb:1990]: US DOE Prototypes (or other similar archetypes) typically do not hold instances of building envelope surfaces that differ in tilt and exposure (within the same model), i.e. with different surface air film resistances. This new tests breaks down uprating/derating steps when processing 3 different roof/ceiling surfaces:

- 1x flat, outdoor-facing roof
- 1x insulated (_interzone_) attic floor/ceiling
- 1x sloped, outdoor-facing roof

Checking if the solution ensures a final, area-weighted roof Ut meeting NECB 2025 targets.

File: spec/tbd_tests_spec.rb [hunk 10]
```diff
@@ -1945,6 +1944,184 @@
...
+    m3 = TBD.derate(slpe.nameString, spc3, lc3)
+    expect(m1.nameString).to eq("Roof uprated m tbd")
+    expect(m2.nameString).to eq("Ceiling uprated m tbd")
+    expect(m3.nameString).to eq("Sloped Roof uprated m tbd")
+    expect(m1.thickness.round(4)).to eq(0.1256)
+    expect(m2.thickness.round(4)).to eq(0.1256)
+    expect(m3.thickness.round(4)).to eq(0.1256)
+    expect(m1.thermalConductivity.round(4)).to eq(0.0151)
+    expect(m2.thermalConductivity.round(4)).to eq(0.0160)
+    expect(m3.thermalConductivity.round(4)).to eq(0.0170)
```
Comment (brgix) [inline spec/tbd_tests_spec.rb:2141]: _Derated_ insulation materials have different k values, as assigned heat loss (from thermal bridging, in W/K) is intentionally spread out unequally.

File: spec/tbd_tests_spec.rb [hunk 11]
```diff
@@ -1945,6 +1944,184 @@
...
+    dR3 = TBD.rsi(lc3, TBD.filmResistances(:roof, slpe.tilt))
+
+    uA  = 0
+    uA += roof.netArea / dR1
+    uA += clng.netArea / dR2
+    uA += slpe.netArea / dR3
+    uT  = uA / area
+
+    # Area-weighted Ut factors meet NECB requirement.
+    expect(uT.round(3)).to eq(u.round(3))
```
Comment (brgix) [inline spec/tbd_tests_spec.rb:2160]: _Uprated_ (then _derated_) roof constructions ultimately meet the NECB 2025 required roof Ut requirement (area-weighted average).

File: spec/tbd_tests_spec.rb [hunk 12]
```diff
@@ -2060,7 +2060,7 @@
     expect(u2.round(5)).to eq(0.07662) # R74.11
     expect(u3.round(5)).to eq(0.07738) # R73.38
 
-    # Matains overall area-weighted (uprated) u0.
+    # Maintains overall area-weighted (uprated) u0.
```
Comment (brgix) [inline spec/tbd_tests_spec.rb:2101]: Typo.

File: lib/tbd/ua.rb [hunk 13]
```diff
@@ -208,9 +211,10 @@ def uprate(model = nil, s = {}, argh = {})
 
       # Collection of one or several constructions to uprate.
       coll = {}
-      op   = g[:op].downcase
+      op   = g[:op]
```
Comment (brgix) [inline lib/tbd/ua.rb:214]: Fixed. This prevented TBD from correctly _uprating_ user-selected constructions.

File: lib/tbd/ua.rb [hunk 14]
```diff
@@ -69,17 +69,17 @@ def uo(id = "", lc = nil, area = 0, film = 0, hloss = 0, ut = 0)
...
 
     # Uprate to counter heat loss from thermal bridging.
     u  = 1 / r
-    u -= hloss / area
+    u -= (hloss / area)
 
     # Adjust if beyond admissible range.
     if u < UMIN
-      negative("#{id}: new Uo", mth, WRN)
+      negative("#{id}: new Uo", mth, INF)
```
Comment (brgix) [inline lib/tbd/ua.rb:82]: Lower-level issues are tagged as _informative_, leaving the higher-level issue raised as a _warning_.

File: lib/tbd/ua.rb [hunk 15]
```diff
@@ -92,8 +92,10 @@ def uo(id = "", lc = nil, area = 0, film = 0, hloss = 0, ut = 0)
       m = m.get.clone(model).to_MasslessOpaqueMaterial.get
       m.setName("#{id} uprated")
 
-      r    = RMIN               if r < RMIN
-      loss = (u - 1 / r) * area if r < RMIN
+      if r < RMIN
+        r    = RMIN
```
Comment (brgix) [inline lib/tbd/ua.rb:96]: Better.

File: lib/tbd/ua.rb [hunk 16]
```diff
@@ -273,11 +277,11 @@ def uprate(model = nil, s = {}, argh = {})
           next
         end
 
-        coll[id] = {}
+        lc = lc.get
```
Comment (brgix) [inline lib/tbd/ua.rb:280]: Fixed. Also prevented TBD from correctly uprating individual constructions.

File: lib/tbd/ua.rb [hunk 17]
```diff
@@ -360,11 +364,16 @@ def uprate(model = nil, s = {}, argh = {})
           col[:s].values.each do |item|
             col[:hloss] += item[:h]
             col[:area ] += item[:a]
-            col[:fA   ] += item[:a] / item[:f]
+            col[:fA   ] += item[:a] / item[:f] unless item[:f] < 0
+          end
+
+          if col[:area] < TOL
+            empty("#{id} area", mth, WRN)
+            next
```
Comment (brgix) [inline lib/tbd/ua.rb:372]: Should never occur - expected to be caught much earlier in the solution. Adding as a precaution (i.e. avoids _divide by zero_ crashes). Multiple fixes in this commit to (finally) catch zero'ed denominators.

File: lib/tbd/psi.rb [hunk 18]
```diff
@@ -1424,8 +1424,10 @@ def derate(id = "", s = {}, lc = nil)
       m  = m.clone(model).to_MasslessOpaqueMaterial.get
       m.setName("#{id} #{up}m tbd")
 
-      de_r = RMIN                    unless de_r > RMIN
-      loss = (de_u - 1 / de_r) * net unless de_r > RMIN
+      if de_r < RMIN
+        de_r = RMIN
+        loss = (de_u - 1 / de_r) * net
```
Comment (brgix) [inline lib/tbd/psi.rb:1429]: Better.

File: lib/tbd/psi.rb [hunk 19]
```diff
@@ -3209,8 +3202,8 @@ def exit(runner = nil, argh = {})
 
       uo     = format("%.3f", g[:uo])
       ut     = format("%.3f", g[:ut])
-      output = "An initial #{label.to_s} Uo of #{uo} W/m2•K is required to " \
-               "achieve an overall Ut of #{ut} W/m2•K for #{g[:op]}"
+      output = "An area-weighted #{label.to_s} Uo of #{uo} W/m2•K is " \
```
Comment (brgix) [inline lib/tbd/psi.rb:3205]: TBD's initial _uprating_ solution would glob all (e.g. wall) constructions together and attempt to _uprate_ a single common wall construction. This offered a few advantages:

- more permissive, accommodating
- easier to calculate
- easier to communicate to users

The downside was that properties (like thermal inertia, or lack of) of individual constructions would be reset by the method.

----

The revised approach maintains user-defined properties of non-insulating layers of individual up/de-rated constructions. The downside is that TBD may now fail to _uprate_ one wall construction in a model (given the extent of heat loss from thermal bridging vs exposed wall area), while succeeding for two others. If TBD fails to _uprate_ one construction, it raises a warning and then defaults the construction Uo to 0.01 W/m2.K. However, the reported Uo (e.g. for all walls) is now an area-weighted Uo (mixing both successful and failed _uprated_ constructions). This may be confusing at first - proceed with caution.

File: spec/tbd_tests_spec.rb [hunk 20]
```diff
@@ -206,6 +205,45 @@
     roof_edges.each { |edg| expect(edg[:surfaces].size).to eq(2) }
   end
 
+  it "can uprate selected construction" do
+    translator = OpenStudio::OSVersion::VersionTranslator.new
+    TBD.clean!
+
+    name  = "Typical Insulated Metal Building Wall R-11.9 1"
+    file  = File.join(__dir__, "files/osms/in/warehouse.osm")
```
Comment (brgix) [inline spec/tbd_tests_spec.rb:213]: New test: a single wall construction selected for _uprating_, reserved for the small corner office space in the US DOE prototype Warehouse.

Image

File: spec/tbd_tests_spec.rb [hunk 21]
```diff
@@ -206,6 +205,45 @@
...
+    lc = model.getLayeredConstructionByName(name)
+    expect(lc).to_not be_empty
+    lc = lc.get
+    expect(lc.getNetArea.round).to eq(126)
+
+    argh                = {}
+    argh[:option      ] = "spandrel HP (BETBG)"
+    argh[:uprate_walls] = true
+    argh[:wall_option ] = name
+    argh[:wall_ut     ] = 0.210 # NECB CZ7 2017 (RSi 4.76 / R27)
```
Comment (brgix) [inline spec/tbd_tests_spec.rb:228]: TBD fails to _uprate_ the selected construction:

- modest floor-to-floor height
- important fenestrated areas
- i.e. small net area
- linear thermal bridging heat loss too great
- NECB 2017-2025 targets too ambitious

TBD no longer spreads out extra heat loss from "office" thermal bridging amongst other walls constructions (i.e. "fine" and "bulk" storage walls) as before.

----

OpenStudio feedback:

Image

----

The issue is raised in red (i.e. "Unable to completely uprate ..."), yet TBD will nonetheless proceed with the uprating/derating steps (with a reset construction Uo of 0.010 W/m2.K). As suggested in TBD's Guide section on [Uprating](https://rd2.github.io/tbd/pages/ut.html):

> We strongly recommend to first investigate this feature while relying on [Apply Measures Now](https://rd2.github.io/tbd/pages/launch.html#apply-measures-now) feedback (ideally UNCHECKing the Alter OpenStudio model option), to get a sense of how significant the uprating calculations may end up altering your initial designs. In some cases, EnergyPlus’ CTF convergence calculations may even fail with very thick (uprated) insulation layers !

File: lib/tbd/ua.rb [hunk 22]
```diff
@@ -454,7 +454,7 @@ def qc33(s = {}, sets = nil, spts = true)
       next unless surface[:deratable]
       next unless surface.key?(:type)
 
-      heating = -50     if spts
+      heating = -24     if spts
```
Comment (brgix) [inline lib/tbd/ua.rb]: Avoids a potential divide by zero crash. Not expecting many thermostats with heating setpoints that low anyways. Rare cases like frozen vaccine storage (-15° down to -80°C), maybe.
